client/src/components/modal-forms/BulkAllocationForm.js

```diff
@@ -0,0 +1,164 @@
+import React, { useState } from 'react';
+import { Button, Dialog } from '../generic';
+import { Box, Typography } from '@material-ui/core';
+import Alert from '@material-ui/lab/Alert';
+import { makeStyles } from '@material-ui/core/styles';
+import { BulkAllocationSchema, ToastStatus } from '../../constants';
+import { RenderTextField, RenderSelectField, RenderCheckbox } from '../fields';
+import { Field, Formik, Form as FormikForm } from 'formik';
+import { dateToTextString } from '../../utils/date';
+import { bulkAllocation } from '../../services/allocations';
+import { useToast } from '../../hooks';
+
+const useStyles = makeStyles(() => ({
+  formButton: {
+    maxWidth: '200px',
+  },
+  formRow: {
+    gap: '25px',
+  },
+  list: {
+    maxHeight: '100px',
+    overflow: 'auto',
+    overflowX: 'hidden',
+  },
+}));
+
+export const createPhaseDropdown = (phases) => {
+  return phases
+    .sort((a, b) => new Date(b.end_date) - new Date(a.end_date))
+    .map((phase) => ({
+      value: phase.id,
+      label: `${phase.name} - ${dateToTextString(phase.start_date)} -
+    ${dateToTextString(phase.end_date)}`,
+    }));
+};
+
+export const BulkAllocationForm = ({ onClose, afterSubmit, open, sites, phases = [] }) => {
+  const { openToast } = useToast();
+  const classes = useStyles();
+  const [existingAllocations, setExistingAllocations] = useState([]);
+  const siteIds = sites.map(({ id }) => id);
+  const initialValues = {
+    allocation: '',
+    phase_id: '',
+    existingAllocations: false,
+    acknowledgement: false,
+  };
+
+  const handleSetAllocation = (phaseId, setValue) => {
+    // get all allocations associated to the selected phase
+    // filter the allocations and return all that are associated to the sites selected.
+    const allocations = phases
+      .find(({ id }) => id === phaseId)
+      ?.allocations.filter(({ site_id }) => siteIds.includes(site_id))
+      .map(({ allocation, phase_id, site_id }) => ({
+        allocation,
+        phase_id,
+        site_id,
+        siteName: sites.find((site) => site.id === site_id).siteName,
+      }));
+
+    setExistingAllocations(allocations);
```

Comment:
Not sure I understand why this needs to be in state AND in the form values? Is it in values because it's being submitted as a payload on submit?

Reply:
The `valuesexistingAllocations` is just a boolean, There didn't seem to be a need to have the entire array in the payload for the yup validation. the state variable `existingAllocations` is being used to display the site/allocations in the alert message.